CDCgov/prime-reportstream
# Updating Authorization to check for Receivers

NOTE: This PR causes the CI server2serverauth test to fail in the pipeline.  We could use some assistance from the RS team in coming up with a solution for this that works for all scenarios


This PR ...

As part of the intermediary effort we need to update the authorization within report stream to check for not only a sender but also a receiver. This would allow the History API to be called by the receivers from Report Stream. It also ensures the intermediary's partners are authorized by report stream when attempting to call the pass through function for metadata.  (NOTE: This function is a work in progress).  This check will also impact the `waters/report/{id}/facilities` endpoint. 

Test Steps:
1.  Add an organization with a sender
2. Add a different organization with a sender AND a receiver
3. Ensure that a message from the sender in step 1 goes to the receiver in step 2
4. As the sender in step 2 attempt to authenticate to report stream and call the history API with the submission ID from step 3 using a generated auth token from the token endpoint

## Changes
- *Include a comprehensive list of changes in this PR*
- *(For web UI changes) Include screenshots/video of changes*

## Checklist

### Testing

### Process

## Linked Issues
- https://github.com/CDCgov/trusted-intermediary/issues/690

## To Be Done
-  Future PR for adding azure function to call the intermediary.  This PR is a standalone change though

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR include changes in authentication and/or authorization of existing endpoints?
  Yes


If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity) Didn't make changes to anything regarding authenticity
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_ Not Applicable
    - **T**ampering (influence or sabotage the integrity of information, data, or system) None, the endpoints in question are read only
    - **R**epudiation (the ability to dispute the origin or originator of an action) We expanded the authorization to include senders and receivers for a given message, this is a change that stays within the existing security framework and should not impact the origin or originator of a given action 
    - **I**nformation disclosure (data made available to entities who should not have it)
      - N/A Data is only being returned to a listed sender or receiver for a given ID as part of this change
    - **D**enial of service (make a resource unavailable)
      - N/A
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
      - N/A
- Have you ensured logging does not contain sensitive data?
  - Yes
- Have you received any additional approvals needed for this change?
  - No
